arch: Add exception context escape trampolines ctx creation

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [ ] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [ ] Tested your changes against relevant architectures and platforms;
 - [ ] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Implement the ability to jump from exception handling context to an
I/O-able context from which one can return to the context that the
exception handler was supposed to return to.

Essentially by creating a return-chain into a function that creates
room on a given stack to store execution environment
(struct ukarch_execenv) which would then return to a custom entry which
would then return to a routine that would restore the aforementioned
execution environment, one is able to freely do I/O, use any registers.

The caller of ukarch_ctx_init_ehtrampoX API's will be given a context
that is guaranteed to run with IRQ's enabled and from which, upon
return, execution from the point the exception handler was supposed
to return to is resumed.
```
        ┌───────────────────┐
▲       │                   │
│       │      struct       │
│       │      execenv      │
│       │                   │
│       │                   │
│       ├───────────────────┤
│       │                   │
│       │  execenv_load(ee) │
│       │                   │
│       ├───────────────────┤
│       │ entry(ee, arg0,   │
│       │       arg1…) for  │
│       │    ehtrampo0,1,2… │
│       │    variants       │
│       │                   │
│       │                   │
│       ├───────────────────┤
│       │                   │
│       │ ehtrampo_.._store │
exec    │                   │
order   │                   │
        └───────────────────┘
```

mogasergiu committed:
arch: Add ctx noreturn jump function

Implement a simple switch to ctx stack and jump to ctx instruction
pointer function.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu commented on arch/x86/ctx.c:
```diff
@@ -286,3 +286,299 @@ void ukarch_ctx_init_entry6(struct ukarch_ctx *ctx,
+	 * we have to push something here that returns into the above return
+	 * chain.
+	 */
+	sp = ukarch_rstack_push(sp, (long)_ctx_x86_call0);
```
Note to self: Write `ehtrampo_ectx_and_sysctx_store` in assembly instead so that we don't bother with this ABI alignment stuff.

mogasergiu committed:
arch: Avoid whitespaces between type and variable when casting

Align with current coding style guidelines and avoid using whitespaces
between the type we want to cast to and the variable we want to cast.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
arch: Add 3 argument ctx creation functions

Replicate the ctx creation functionality for the cases in which
one may want to create a context with a function that requires 3
arguments.

Checkpatch-Ignore: AVOID_EXTERNS
Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
arch: Add exception context escape trampolines ctx creation

Implement the ability to jump from exception handling context to an
a context where one can take exceptions or yield the current thread and
from which one can return to the context that the
exception handler was supposed to return to.

By calling `ukarch_ctx_init_ehtrampo`, caller passes a usable stack
pointer as one of the arguments where the general purpose registers
along with what else makes the execution envirionments will be stored
as well as the argument passed by the caller. The caller passed
callback will be guaranteed to be in a flexible context where it can do
everything it wants given the guarantee that upon exit, it will return
safely to the context that generated the exception in the first place.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented on include/uk/arch/ctx.h:
```diff
@@ -340,6 +340,50 @@ void ukarch_ctx_init_entry2(struct ukarch_ctx *ctx,
+typedef void (*ukarch_ehtrampo_entry)(struct ukarch_execenv *, long);
+
+/**
+ * Initializes a context struct with stack pointer and
```
> Initializes a context struct with stack pointer and entrance function to trampoline to from an exception handler to an I/O-able context.

I had to read the implementation to understand the documentation. It should be the other way around 🙃 

Consider moving the sentence in line 356 to the top, as it's generally more useful to explain what it is before you explain how it works. I would even go as far as moving the "how" to the comments in the code.

Another thing that is missing, and again more imortant on how it works, is how to use it. So documentation should also state that you are expected to call `ukarch_ectx_jmp()`. Saying that, it would be worth considering whether the init and jump operations should be merged to a single call.

PS: "Entry" is more commonly used than "entrance" in this context.

mogasergiu replied:
> it would be worth considering whether the init and jump operations should be merged to a single call.

Thought about this as well. I did it like this the first time to align it with the other `ukarch_ctx_init_*` functions. I am trying to come up with a fitting name for something that implies it does both.

michpappas commented on arch/x86/ctx.c:
```diff
@@ -189,3 +190,54 @@ void ukarch_ctx_init_entry3(struct ukarch_ctx *ctx,
+	ukarch_sysctx_store(&ee->sysctx);
+
+	/*
+	 * NOTE: This is very risky to do in ISR context as the compiler
```
At first I thought you were trying to prevent the compiler from using `memcpy()`. Proposed phrasing, short and to the point:

> NOTE: Order is important. We save the registers after ectx . As the compiler could implement this using memcpy(), we run the risk of corrupting the ectx before we had a chance to save it.

michpappas commented on include/uk/arch/ctx.h:
```diff
@@ -340,6 +340,50 @@ void ukarch_ctx_init_entry2(struct ukarch_ctx *ctx,
+ *
+ * @param ctx
+ *   Reference to context to initialize
+ * @param sp
```
Perhaps you should clarify the purpose of that stack. Perhaps something along the lines of:
> Should point to the stack the target context will execute on. Unless you have a reason, most of the time that would be the auxilliary stack.

On second thought, one idea that could make the caller's life easier is to have el trampo use the aux stack by default, unless the caller passes a non-NULL pointer. Since both el trampo and the auxiliary stack will be part of the same library, that should not be too obtrusive.

Another thing I noticed is that el trampo pushes into the stack. Perhaps it's worth to mention that, in case the caller provides a pointer to space-confined stack.
 
In addition to the above, you might want to switch the order of `sp` under that of `r` in the docstring, to match the function's order of  parameters.

mogasergiu replied:
> On second thought, one idea that could make the caller's life easier is to have el trampo use the aux stack by default, unless the caller passes a non-NULL pointer. Since both el trampo and the auxiliary stack will be part of the same library, that should not be too obtrusive.

That would be nice. The problem is the way it is used may differ drastically depending on what the target context may be (e.g. user app handler or nested exception handler)

michpappas replied:
Okay I see, let's drop it then 👍🏼

michpappas commented on include/uk/arch/ctx.h:
```diff
@@ -340,6 +340,50 @@ void ukarch_ctx_init_entry2(struct ukarch_ctx *ctx,
 void ukarch_ctx_switch(struct ukarch_ctx *store, struct ukarch_ctx *load);
 
+/*
+ * (Full) Execution environment functions must return into environment restoring
```
With no parameter names and no documentation, this is quite cryptic. Would be worth moving the description from `ukarch_ctx_init_ehtrampo()` here. Also giving names to the parameters makes the reader's life easier, and IMO in a public API it's a must. Plus you can also use these names in docstrings.

michpappas commented on arch/x86/ctx.c:
```diff
@@ -189,3 +190,54 @@ void ukarch_ctx_init_entry3(struct ukarch_ctx *ctx,
+	UK_ASSERT(!(sp & UKARCH_SP_ALIGN_MASK)); /* sp properly aligned? */
+
+	/* We re-align the stack anyway just to be sure */
+	sp = ALIGN_DOWN(sp, UKARCH_EXECENV_END_ALIGN);
```
If it's stated as requirement already, then pehaps an assert is more suitable. Either that or not make it a requirement and do the align internally.

michpappas commented:
In addition to the above, it would be useful to provide in the cover letter and the commit message that implements the trampoline logic, what is the motivation for this, i.e. problem does it solve.

mogasergiu commented:
Pushed rebase

mogasergiu commented:
Pushed updated addressing comments

michpappas commented on include/uk/arch/ctx.h:
```diff
@@ -339,6 +340,59 @@ void ukarch_ctx_init_entry2(struct ukarch_ctx *ctx,
+ * Initializes a context struct with stack pointer and
+ * entry function to trampoline to from an exception handler
+ * into an I/O-able context.
+ * Does not guarantee any register states when entering the user-passed
```
This can be also moved to the docstring of `ukarch_ehtrampo_entry()`.

michpappas commented on include/uk/arch/ctx.h:
```diff
@@ -339,6 +340,59 @@ void ukarch_ctx_init_entry2(struct ukarch_ctx *ctx,
+
+/**
+ * Initializes a context struct with stack pointer and
+ * entry function to trampoline to from an exception handler
```
In this line, s/to from/from/

However, as stated on the previous comment, perhaps it's better to focus on the information a caller would need: what it does and how to use it (as opposed to how it works). Notice also how you repeat yourself in two subsequent paragraphs:

>  * Initializes a context struct ... to trampoline to from an exception handler
>  * into an I/O-able context.

and on the very next paragraph again the same statement:

>  * This function is most useful when wanting to go from an exception
> * handler to an I/O-able context while being able to return to the
> * place that the exception handler was supposed to return to.

You also still don't tell the caller how to use it. I would propose something along the lines of:

> Initializes a context that allows  jumping from an exception handling context to a caller-defined function in an  I/O-able context.  Returning from the caller-defined function resumes execution from the place that the exception handler would resume from.
>
> After calling `ukarch_ehtrampo_init()` use `ukarch_ctx_jmp()` to jump to the target context.
> 
> NOTE: This function may return with a tainted extended context.

michpappas commented on include/uk/arch/ctx.h:
```diff
@@ -339,6 +340,59 @@ void ukarch_ctx_init_entry2(struct ukarch_ctx *ctx,
+
+/**
+ * Function that can be executed as if outside exception context. After the
+ * function ends, execution will restore the state stored in the execenv
```
after the function *returns*, execution will *resume*

michpappas commented on include/uk/arch/ctx.h:
```diff
@@ -339,6 +340,59 @@ void ukarch_ctx_init_entry2(struct ukarch_ctx *ctx,
+ * @param execenv
+ *   Pointer to the execution environment that the function will return to
+ * @param arg
+ *   Custom user-passed argument
```
User-defined argument

michpappas commented on include/uk/arch/ctx.h:
```diff
@@ -339,6 +340,59 @@ void ukarch_ctx_init_entry2(struct ukarch_ctx *ctx,
+void ukarch_ctx_jump(struct ukarch_ctx *ctx) __noreturn;
+
+/**
+ * Function that can be executed as if outside exception context. After the
```
To keep consistent terminology:
> executed in the I/O-able context

michpappas reviewed (changes requested):
A few more improvements on the API documentation.

mogasergiu commented:
Pushed addressing comments
+
Removed the "I/O-able" term and replaced it with "a context where one can take exceptions or yield the current thread"
In earlier versions, the API would always implicitly enable interrupts, making it true that one can do even interrupt-based I/O. But, in current versions, the caller or the called function would have to enable interrupts manually, if there weren't already and if they wanted to.
The reason the API no longer implicitly enables interrupts is because in a scenario where the caller is already in an interrupt-enabled context, the API would needlessly re-enable them.
Therefore now, the API "teleports" the caller to a context where it can taint whatever register context it wants, but the state of interrupt flags is the state execution had before jumping to aforementioned context.

michpappas reviewed (approved):
Thanks!

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

andreittr reviewed (approved):
Reviewed-by: Andrei Tatar <ttr@unikraft.io>

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
